lib/srv/git/github.go

```diff
@@ -30,35 +34,102 @@ import (
 	"google.golang.org/grpc"
 	"google.golang.org/protobuf/types/known/durationpb"
 
+	"github.com/gravitational/teleport"
 	integrationv1 "github.com/gravitational/teleport/api/gen/proto/go/teleport/integration/v1"
 	"github.com/gravitational/teleport/api/types"
 	"github.com/gravitational/teleport/lib/cryptosuites"
+	"github.com/gravitational/teleport/lib/defaults"
 	"github.com/gravitational/teleport/lib/sshutils"
 )
 
-// knownGithubDotComFingerprints contains a list of known GitHub fingerprints.
-//
-// https://docs.github.com/en/authentication/keeping-your-account-and-data-secure/githubs-ssh-key-fingerprints
-//
-// TODO(greedy52) these fingerprints can change (e.g. GitHub changed its RSA
-// key in 2023 because of an incident). Instead of hard-coding the values, we
-// should try to periodically (e.g. once per day) poll them from the API.
-var knownGithubDotComFingerprints = []string{
-	"SHA256:uNiVztksCsDhcc0u9e8BujQXVUpKZIDTMczCvj3tD2s",
-	"SHA256:p2QAMXNIC1TJYWeIOttrVc98/R1BUFWu3/LiyKgUfQM",
-	"SHA256:+DiY3wvvV6TuJJhbpZisF/zLDA0zPMSvHdkr4UvCOqU",
+// githubServerKeyManager downloads SSH keys from the GitHub meta API and does a
+// lazy refresh every hour. The keys are used to verify GitHub server when
+// forwarding Git commands to it.
+type githubServerKeyManager struct {
+	mu        sync.Mutex
+	keys      []string
+	lastCheck time.Time
+	etag      string
+
+	clock       clockwork.Clock
+	apiEndpoint string
+	client      *http.Client
+}
+
+func newGitHubServeKeyManager() *githubServerKeyManager {
+	return &githubServerKeyManager{
+		clock:       clockwork.NewRealClock(),
+		apiEndpoint: "https://api.github.com/meta",
+		client: &http.Client{
+			Timeout: defaults.HTTPRequestTimeout,
+		},
+	}
 }
 
-// VerifyGitHubHostKey is an ssh.HostKeyCallback that verifies the host key
-// belongs to "github.com".
-func VerifyGitHubHostKey(_ string, _ net.Addr, key ssh.PublicKey) error {
-	actualFingerprint := ssh.FingerprintSHA256(key)
-	if slices.Contains(knownGithubDotComFingerprints, actualFingerprint) {
+func (m *githubServerKeyManager) check(targetKey ssh.PublicKey) error {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	// Refresh every 24 hours.
+	if m.clock.Now().Sub(m.lastCheck) > time.Hour*24 {
+		m.refreshLocked()
+	}
+
+	// Remove newline from ssh.MarshalAuthorizedKey.
+	key := strings.TrimSpace(string(ssh.MarshalAuthorizedKey(targetKey)))
+	if slices.Contains(m.keys, key) {
```

Comment:
To avoid confusing errors, I'd check that `m.keys` is non-empty as a special case.